### DISCUSSION TITLE
Entity nesting, repeaters, wizards and all that jazzz

### DISCUSSION THREAD

**Discussion by aglavas** [filamentphp/filament#6134] [Help]
Hello, I have a question about my situation. Wondering what I am doing wrong, or how to achieve what I want.
So, I have a few nested resources.

User > Benchmark > Dimension > Question.

So the User has many Benchmarks. And Benchmark has many dimensions. And finally Dimension has many questions.
For simplicity reasons we can ignore first level of nesting (User owning many Benchmarks).
So then it looks like this:

Benchmark > Dimension > Question

Here is the diagram representing situation in DB. 

![image](https://user-images.githubusercontent.com/9666109/230248789-e47cff2f-ea61-4433-bb4e-8228fd6a574f.png)

In diagram there is also one additional table that maps everything.

What is also important, is that these 3 entities (which are related in parent-child relation between them) are created in multi step process. I have picked a Wizard to do this (since it seemed the most convenient for the job).

The relevant code:

App\Filament\Resources\BenchmarkResource\Pages\CreateBenchmark

```
namespace App\Filament\Resources\BenchmarkResource\Pages;

use App\Filament\Resources\BenchmarkResource;
use Filament\Forms\Components\Card;
use Filament\Forms\Components\Wizard\Step;
use Filament\Resources\Pages\CreateRecord;

class CreateBenchmark extends CreateRecord
{
    use CreateRecord\Concerns\HasWizard;

    protected static string $resource = BenchmarkResource::class;

    protected function getSteps(): array
    {
        return [
            Step::make('Create Benchmark')
                ->schema([
                    Card::make(BenchmarkResource::getFormSchema('benchmark')),
                ]),
            Step::make('Create Dimensions')
                ->schema([
                    Card::make(BenchmarkResource::getFormSchema('dimensions')),
                ]),
            Step::make('Create Questions')
                ->schema([
                    Card::make(BenchmarkResource::getFormSchema('questions')),
                ]),
        ];
    }
}

```

App\Filament\Resources\BenchmarkResource

```
namespace App\Filament\Resources;

use App\Filament\Resources\BenchmarkResource\Pages;
use App\Filament\Resources\BenchmarkResource\RelationManagers;
use App\Helpers\CountryHelper;
use App\Models\Benchmark;
use Filament\Forms;
use Filament\Forms\Components\Select;
use Filament\Forms\Components\TagsInput;
use Filament\Forms\Components\ViewField;
use Filament\Resources\Resource;
use Filament\Resources\Table;
use Filament\Tables;

class BenchmarkResource extends Resource
{
    protected static ?string $model = Benchmark::class;

    protected static ?string $navigationIcon = 'heroicon-o-collection'; 

    public static function getFormSchema($key): array
    {
        $schemaArray = [
            'benchmark' => [
                Forms\Components\TextInput::make('account_id')->default(1)
                    ->hidden(),
                Forms\Components\TextInput::make('title')
                    ->reactive()
                    ->label(__('fields.benchmark.title'))
                    ->required()
                    ->maxLength(255),
                Forms\Components\Textarea::make('description')
                    ->reactive()
                    ->label(__('fields.benchmark.description'))
                    ->helperText(__('fields.benchmark.description.text'))
                    ->required(),
                TagsInput::make('audience')
                    ->reactive()
                    ->label(__('fields.benchmark.audience'))
                    ->helperText(__('fields.benchmark.audience.text'))
                    ->suggestions([
                        'Head of marketing',
                        'VP Sales',
                        'Human Resource',
                        'CTO',
                        'Developers',
                    ])
                    ->required(),
                Select::make('countries')
                    ->reactive()
                    ->multiple()
                    ->options(CountryHelper::getCountryList())
                    ->label(__('fields.benchmark.countries'))
                    ->id('countries')
                    ->helperText(__('fields.benchmark.countries.text'))
                    ->required(),
                Select::make('language')
                    ->reactive()
                    ->label(__('fields.benchmark.language'))
                    ->id('language')
                    ->helperText(__('fields.benchmark.language.text'))
                    ->options([
                        'English' => 'English',
                    ])
                    ->required(),
                Forms\Components\Toggle::make('active')
                    ->reactive()
                    ->required(),
            ],
            'dimensions' => [
                ViewField::make('dimensions')->view('filament.view.dimensions'),
                Forms\Components\Repeater::make('dimensions')
                    ->relationship()
                    ->schema([
                        Forms\Components\TextInput::make('title')
                            ->reactive()
                            ->label(__('fields.benchmark.dimensions.title'))
                            ->required()
                            ->maxLength(255),

                        Forms\Components\Textarea::make('description')
                            ->reactive()
                            ->label(__('fields.benchmark.dimensions.description'))
                            ->helperText(__('fields.benchmark.dimensions.description.text'))
                            ->required(),
                    ])
                    ->defaultItems(2)
                    ->disableItemCreation()
                    ->disableItemDeletion()
                    ->disableItemMovement()
                    ->columns(1)
                    ->required(),
            ],
            'questions' => [
                Forms\Components\Repeater::make('questions')
                    ->relationship()
                    ->schema([
                        Forms\Components\TextInput::make('title')
                            ->reactive()
                            ->label(__('fields.benchmark.dimensions.title'))
                            ->required()
                            ->maxLength(255),

                        Forms\Components\Textarea::make('description')
                            ->reactive()
                            ->label(__('fields.benchmark.dimensions.description'))
                            ->helperText(__('fields.benchmark.dimensions.description.text'))
                            ->required(),
                    ])
                    ->defaultItems(function ($livewire, \Closure $get, \Closure $set, string $context) {
                        return count($livewire->data['dimensions']);
                    })
                    ->columns(1)
                    ->required(),
            ],
        ];

        return $schemaArray[$key];
    }
}
```

This setup would achieve creation of Benchmark entity and multiple related Dimension entities using Repeater in Wizzard.
However it will fail on final step, since it will expect questions relationship on Benchmark model.

I tried to declare HasManyThrough relationship (from Benchmark > (over Dimension) > to Question), but it was still complaining about Repeater (which doesn't support HasManyThrough) relationship.

Is there a way to do it? But that I still have multi step form definition of resources?
Any suggestions?

**Comment by danharrin** [comment]
You've provided lots of details but I'm not entirely sure I grasp the concept. I hope this helps though - https://filamentphp.com/docs/2.x/forms/layout#saving-data-to-relationships

You can wrap any field, including a repeater, in a layout component like Fieldset or Grid or Group and teleport those fields into a relationship, so that might be useful

**Comment by aglavas** [comment]
Let me try to rephrase it.
Let's imagine we have 3 entities. A, B and C in parent-child relationship (hasMany).
So:
A hasMany B (reference would be A_id field in database in B's table) and 
B hasMany C (reference would be B_id field in database in C's table).

Can this be done and created through wizard (or tab) flow?

I know that you can declare a schema in which you create entity A and it's children by creating a Repeater (with B's fields) inside entity A. Similar how here is done - https://demo.filamentphp.com/shop/orders/create ( https://github.com/filamentphp/demo/blob/903011d7d75b648d13f4c66f8878116ddd1642e6/app/Filament/Resources/Shop/OrderResource.php#L202 & https://github.com/filamentphp/demo/blob/main/app/Filament/Resources/Shop/OrderResource/Pages/CreateOrder.php).

My question is can you on top of that add a repeater (creation of hasMany relationship) to each entity B (or each entity OrderItems from https://demo.filamentphp.com example)?

**Reply by danharrin** [reply]
You can nest two relationship repeaters within each other to create a nested structure, yes

**Reply by orbitsdev** [reply]
> You can nest two relationship repeaters within each other to create a nested structure, yes

Sir danharrin.. I am using filament 3

Repeater..

Let say I have.. relationship models of

- Appointment hat has many Patient
 -Patient belongs to animal..

Now when i diplay the patience
Using.. repeater it works. It loop 


But when I want to access.. the animal details let say name it is not showing


I am trying to access it like this



---patients (assume it is repeater)
  ---- animal.name (inside repeater)

Is this wrong?

I tried also this

----patients
    ----patients.animal.name

It doesn't show..

Is Repeater cannot access child nested relationship?

I hope.. in repeater docs.. we can see few example how we can access nested child.. using repeater

**Comment by aglavas** [comment]
You also got me thinking with other part of your reply. 
Can you maybe explain what effect layout components would have in terms of relationship?.

E.g. Is there any difference between:

Repeater::make('qualifications')
    ->relationship()
    ->schema([
        // ...
    ])
    
    and
    
Fieldset::make('Qualifications')
    ->relationship('qualifications')
    ->schema([
        Repeater::make('qualifications')
        ->relationship()
        ->schema([
            // ...
        ])
    ])

?

**Comment by danswiser** [comment]
I believe I have the same issue/questions. Did you ever find a solution @aglavas?

Does it make more sense to save each step of the wizard to database when clicking next and is that possible?

In simplest terms, nesting repeaters is probably a solution that would work but for usability, I'm trying to move the nested repeater to a separate step in the wizard and put the parent repeater items in a select option.

**Reply by guilhermemelgarejo** [reply]
I am with the same question, what solution did you find?

**Reply by aglavas** [reply]
@danswiser @guilhermemelgarejo In the end, I have re-coded the wizard custom. In the end it was saving the each step to DB + walking over the wizard to previous and next steps... Re-implementing that component was horrible experience.

So the part with repeaters look like this:


public static function getDimensionFormSchema($mode)

    {
        return [
            ViewField::make('dimensions')
                ->view('filament.view.dimensions'),
            Repeater::make('dimensions')
                ->schema(DimensionResource::getFormSchema($mode))
                ->defaultItems(6)
                ->disableItemCreation()
                ->disableItemDeletion()
                ->disableItemMovement()
                ->afterStateHydrated(function (\Closure $set, \Closure $get, Repeater $component, $state) {
                    $stateCount = count($state);
                    while ($stateCount < 6) {
                        $newUuid = (string) Str::uuid();
                        $livewire = $component->getLivewire();
                        data_set($livewire, "data.dimension.dimensions.{$newUuid}", []);
                        $component->getChildComponentContainers()[$newUuid]->fill();
                        $component->collapsed(false, shouldMakeComponentCollapsible: false);
                        $stateCount++;
                    }
                })->columns(1)
                ->rules([
                    function () use ($mode) {
                        return function (string $attribute, $data, \Closure $fail) use ($mode) {
                            $failed = false;
                            if ($mode === 'create') {                                
                                $validationData = [];
                                foreach ($data as $repeaterKey => $item) {
                                    if (!empty($item['title'])) {
                                        $validationData[$repeaterKey] = $item;
                                    }
                                }

                                if (count($validationData) == 0) {
                                    $failed = true;
                                    $fail("At least one dimension title should be filled");
                                }

                                if (!$failed) {
                                    $repeaterData = ['repeaterData' => $validationData];
                                    try {
                                        Validator::validate($repeaterData, [
                                            'repeaterData' => 'required|array|arrayMinFilled',
                                            'repeaterData.*.title' => 'distinct',
                                        ]);
                                    } catch (ValidationException $exception) {
                                        $fail('Dimension title should be unique.');
                                    }
                                }
                            } elseif ($mode === 'edit') {
                                $editValidationData = [];
                                $createValidationData = [];
                                foreach ($data as $repeaterKey => $item) {
                                    $isEditData = is_int($repeaterKey);
                                    if ($isEditData) {
                                        $editValidationData[$repeaterKey] = $item;
                                    } else {
                                        $createValidationData[$repeaterKey] = $item;
                                    }
                                }

                                $repeaterData = ['repeaterData' => $editValidationData];

                                try {
                                    Validator::validate($repeaterData, [
                                        'repeaterData' => 'required|array',
                                        'repeaterData.*.title' => 'required|distinct',
                                    ]);
                                } catch (ValidationException $exception) {
                                    $fail($exception->getMessage());
                                    $failed = true;
                                }

                                if (!$failed) {
                                    $validationData = [];
                                    foreach ($createValidationData as $repeaterKey => $item) {
                                        if (!empty($item['title'])) {
                                            $validationData[$repeaterKey] = $item;
                                        }
                                    }

                                    if (count($validationData) > 0) {
                                        $finalValidationData = array_merge($editValidationData, $validationData);
                                        $repeaterData = ['repeaterData' => $finalValidationData];
                                        try {
                                            Validator::validate($repeaterData, [
                                                'repeaterData' => 'required|array',
                                                'repeaterData.*.title' => 'required|distinct',
                                            ]);
                                        } catch (ValidationException $exception) {
                                            $fail('Dimension title should be unique.');
                                        }
                                    }
                                }
                            }
                        };
                    },
                ]),
        ];
    }

And:

`   
 public static function getQuestionFormSchema(bool $edit = false)
 
   {
        return [

            Repeater::make('question_form')->schema([
               Accordions::make('question_form.accordion')
                   ->label(function (CreateBenchmark $benchmark, Component $component, \Closure $get, \Closure $set, string $context, array $state) {
                        /** @var ComponentContainer $container */
                        $container = $component->getContainer();
                        $livewire = $container->getLivewire();
                        $dimensionArray = $livewire->dimensionTitleArray;
                        $questionForm = $livewire->data['question']['question_form'];
                        $statePath = $component->getStatePath();
                        $questionFormKeys = array_flip(array_keys($questionForm));
                        $statePathExploded = explode('.', $statePath);
                        $repeaterKey = end($statePathExploded);
                        $nthKey = $questionFormKeys[$repeaterKey];
                        $dimensionsCollection = collect($dimensionArray);
                        $dimensionValuesCollection = $dimensionsCollection->values();

                        if (!count($dimensionValuesCollection)) {
                             return '';
                        }

                        $dimensionTitle = $dimensionValuesCollection->get($nthKey);
                        return $dimensionTitle;
                    })
                    ->accordions([
                        Accordions\Accordion::make("")
                            ->schema([
                                Repeater::make('questions')
                                    ->schema(QuestionResource::getFormSchema($edit))
                                    ->defaultItems(1)
                                    ->disableItemDeletion(function ($state, \Closure $get ) {
                                        $stateCount = count($state);

                                        if ($stateCount == 1) {
                                            return true;
                                        }

                                        return false;
                                    })
                                    ->disableItemMovement()
                                ])
                            ])->compact()
                        ])->columns(6)
                ->defaultItems(function ($livewire, \Closure $get, \Closure $set, string $context, $state) {
                    if ($livewire->dimensionCount) {
                        return $livewire->dimensionCount;
                    }

                    return 6;
                })
                ->columns(1)
                ->disableItemCreation()
                ->disableItemDeletion()
                ->disableItemMovement()
                ->afterStateHydrated(function (\Closure $set, \Closure $get, Repeater $component, $state) {
                    $liveWire = $component->getLivewire();
                    $currentStateData = $liveWire->data;
                    $filledDimensionCount = 0;
                    $stateCount = count($state);

                    foreach ($currentStateData['dimension']['dimensions'] as $dimensionData) {
                        if (isset($dimensionData['id'])) {
                            $filledDimensionCount++;
                        }
                    }

                    if ($stateCount < $filledDimensionCount) {
                        while ($stateCount < $filledDimensionCount) {
                            $newUuid = (string) Str::uuid();
                            $livewire = $component->getLivewire();
                            data_set($livewire, "data.question.question_form.{$newUuid}.questions", []);
                            $component->getChildComponentContainers()[$newUuid]->fill();
                            $component->collapsed(false, shouldMakeComponentCollapsible: false);
                            $stateCount++;
                        }
                    }
                })
                ->required()
            ];
    }`

Then, in the custom implemented Wizard component I would just reference this structure by doing:


    protected function getBenchmarkForm(): array
    {       
        $schema = ($this->mode == 'edit') ? array_merge(BenchmarkResource::getFormSchema(), [ Toggle::make('active')->required() ]) : BenchmarkResource::getFormSchema();
        return [
            Card::make($schema)
        ]; 
    }
    
    protected function getDimensionForm(): array
    {
        $schema = BenchmarkResource\Pages\CreateBenchmark::getDimensionFormSchema($this->mode);
        return [
            Card::make(
                $schema
            ),
        ];
    }

It does work, but the amount of time invested in this approach which relies on custom re-implementation was just not worth the effort - at least not for me.
